File: CS/WebApplication_GridTest/Templates/CustomEditFormTemplate.cs
using DevExpress.Web.ASPxEditors;
using DevExpress.Web.ASPxGridView;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.UI;
using System.Web.UI.HtmlControls;

namespace WebApplication_GridTest {
    public class CustomEditFormTemplate : ITemplate {
        public void InstantiateIn(Control _container) {
            GridViewEditFormTemplateContainer container = _container as GridViewEditFormTemplateContainer;
            object[] values = (object[])container.Grid.GetRowValues(container.VisibleIndex, new string[] { "Id", "Type", "DataValue" });
            string id = Convert.ToString(values[0]);
            string dataType = Convert.ToString(values[1]);
            string dataValue = Convert.ToString(values[2]);

            ASPxTextBox tb = new ASPxTextBox();
            tb.ID = "id";
            container.Controls.Add(tb);
            tb.Text = values[0].ToString();
            tb.ReadOnly = true;

            tb = new ASPxTextBox();
            tb.ID = "datatype";
            container.Controls.Add(tb);
            tb.Text = values[1].ToString();
            tb.ReadOnly = true;

            switch (dataType) {
                case "Text":
                    ASPxTextBox txt = new ASPxTextBox();
                    txt.ID = "txt";
                    container.Controls.Add(txt);
                    txt.Text = dataValue;
                    break;
                case "DateTime":
                    ASPxDateEdit date = new ASPxDateEdit();
                    date.ID = "date";
                    container.Controls.Add(date);
                    date.Date = Convert.ToDateTime(dataValue);
                    break;
                case "Dictionary1":
                    ASPxComboBox cmb = new ASPxComboBox();
                    cmb.ID = "dict";
                    container.Controls.Add(cmb);
                    cmb.DataSource = DataProvider.GetDictionaryData(DataType.Dictionary1);
                    cmb.TextField = "Data";
                    cmb.ValueField = "Id";
                    cmb.DataBind();
                    cmb.Value = dataValue;
                    break;
                case "Dictionary2":
                    ASPxComboBox cmb2 = new ASPxComboBox();
                    cmb2.ID = "dict";
                    container.Controls.Add(cmb2);
                    cmb2.DataSource = DataProvider.GetDictionaryData(DataType.Dictionary2);
                    cmb2.TextField = "Data";
                    cmb2.ValueField = "Id";
                    cmb2.DataBind();
                    cmb2.Value = dataValue;
                    break;
            }

            ASPxGridViewTemplateReplacement updateBtn = new ASPxGridViewTemplateReplacement();
            updateBtn.ReplacementType = GridViewTemplateReplacementType.EditFormUpdateButton;
            container.Controls.Add(updateBtn);

            ASPxGridViewTemplateReplacement cancelBtn = new ASPxGridViewTemplateReplacement();
            cancelBtn.ReplacementType = GridViewTemplateReplacementType.EditFormCancelButton;
            container.Controls.Add(cancelBtn);
        }
    }
}

File: VB/WebApplication_GridTest/Templates/CustomEditFormTemplate.vb
Imports Microsoft.VisualBasic
Imports DevExpress.Web
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports System.Web.UI
Imports System.Web.UI.HtmlControls

Namespace WebApplication_GridTest
	Public Class CustomEditFormTemplate
		Implements ITemplate
		Public Sub InstantiateIn(ByVal _container As Control) Implements ITemplate.InstantiateIn
			Dim container As GridViewEditFormTemplateContainer = TryCast(_container, GridViewEditFormTemplateContainer)
			Dim values() As Object = CType(container.Grid.GetRowValues(container.VisibleIndex, New String() { "Id", "Type", "DataValue" }), Object())
			Dim id As String = Convert.ToString(values(0))
			Dim dataType As String = Convert.ToString(values(1))
			Dim dataValue As String = Convert.ToString(values(2))

			Dim tb As New ASPxTextBox()
			tb.ID = "id"
			container.Controls.Add(tb)
			tb.Text = values(0).ToString()
			tb.ReadOnly = True

			tb = New ASPxTextBox()
			tb.ID = "datatype"
			container.Controls.Add(tb)
			tb.Text = values(1).ToString()
			tb.ReadOnly = True

			Select Case dataType
				Case "Text"
					Dim txt As New ASPxTextBox()
					txt.ID = "txt"
					container.Controls.Add(txt)
					txt.Text = dataValue
				Case "DateTime"
					Dim [date] As New ASPxDateEdit()
					[date].ID = "date"
					container.Controls.Add([date])
					[date].Date = Convert.ToDateTime(dataValue)
				Case "Dictionary1"
					Dim cmb As New ASPxComboBox()
					cmb.ID = "dict"
					container.Controls.Add(cmb)
					cmb.DataSource = DataProvider.GetDictionaryData(WebApplication_GridTest.DataType.Dictionary1)
					cmb.TextField = "Data"
					cmb.ValueField = "Id"
					cmb.DataBind()
					cmb.Value = dataValue
				Case "Dictionary2"
					Dim cmb2 As New ASPxComboBox()
					cmb2.ID = "dict"
					container.Controls.Add(cmb2)
                    cmb2.DataSource = DataProvider.GetDictionaryData(WebApplication_GridTest.DataType.Dictionary2)
					cmb2.TextField = "Data"
					cmb2.ValueField = "Id"
					cmb2.DataBind()
					cmb2.Value = dataValue
			End Select

			Dim updateBtn As New ASPxGridViewTemplateReplacement()
			updateBtn.ReplacementType = GridViewTemplateReplacementType.EditFormUpdateButton
			container.Controls.Add(updateBtn)

			Dim cancelBtn As New ASPxGridViewTemplateReplacement()
			cancelBtn.ReplacementType = GridViewTemplateReplacementType.EditFormCancelButton
			container.Controls.Add(cancelBtn)
		End Sub
	End Class
End Namespace